### PR TITLE
EI 2028 - Function app terraform module should be extended with Private endpoint support (1 of 2)

### PR DESCRIPTION
**Pull request by swiftnickmoj**
### Jira link (if applicable)
https://tools.hmcts.net/jira/browse/EI-2028

### Change description ###
Modified code to make function apps accessible internally only via private endpoints.

### Checklist

- [x] commit messages are meaningful and follow good commit message guidelines
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -137,7 +164,8 @@ resource "azurerm_windows_function_app" "windows_function" {
   client_certificate_enabled  = var.client_certificate_enabled
   client_certificate_mode     = var.client_certificate_mode
   builtin_logging_enabled     = var.builtin_logging_enabled
-  virtual_network_subnet_id   = length(var.subnet_name) == 0 ? azurerm_subnet.main[0].id : var.subnet_id
+  virtual_network_subnet_id   = length(var.subnet_name) == 0 ? azurerm_subnet.main[0].id : data.azurerm_subnet.main.0.id
+  public_network_access_enabled = false
```
Comment (varun-vijayagopal) [inline main.tf:168]: what will happen if it is a standard sku ? shouldn't this be true in that case ?

File: main.tf [hunk 2]
```diff
@@ -193,3 +221,29 @@ resource "azurerm_app_service_public_certificate" "functionapp" {
...
+# Link Private DNS zone to VNet and disable public address to Function App
+# Private DNS Zone should already exist via cpp-terraform-network
+
+data "azurerm_private_dns_zone" "dns_zone" {
+  name                = "privatelink.azurewebsites.azure.net"
+  resource_group_name = var.dns_resource_group_name
+}
+
+resource "azurerm_private_dns_a_record" "dns_record" {
+  #name                = azurerm_function_app.function.name
```
Comment (varun-vijayagopal) [inline main.tf:234]: shouldn't this be conditional like privat endpoint based on skus ?

File: main.tf [hunk 3]
```diff
@@ -193,3 +221,29 @@ resource "azurerm_app_service_public_certificate" "functionapp" {
...
+  name                = var.asp_os_type == "Linux" ? azurerm_linux_function_app.linux_function.0.name : azurerm_windows_function_app.windows_function.0.name
+  #zone_name           = azurerm_private_dns_zone.rg.name
+  zone_name           = data.azurerm_private_dns_zone.dns_zone.name
+  resource_group_name = var.dns_resource_group_name
+  ttl                 = 300
+  records             = [azurerm_private_endpoint.private_endpoint[0].private_service_connection[0].private_ip_address]
+}
+
+resource "azurerm_private_dns_zone_virtual_network_link" "dns_link" {
+  name                  = "LinkDNSZoneVNet"
```
Comment (varun-vijayagopal) [inline main.tf:244]: shouldn't this be conditional like privat endpoint based on skus ?

Comment (varun-vijayagopal) [reply main.tf:244]: also shouldn't this vnet link creation to be done on cpp-terrafrom-networks repo ?

File: variables.tf [hunk 4]
```diff
@@ -169,3 +180,32 @@ variable "storage_account" {
...
+variable "dns_link" {
+  type        = string
+  description = "Name of DNS link connecting private DNS zone to VNet"
+  default     = ""
+}
+
+variable "dns_resource_group_name" {
+  type = string
+  description = "Name of private DNS zone resource group"
+}
```
Comment (varun-vijayagopal) [inline variables.tf:211]: fix pre-commit

**Review by varun-vijayagopal — CHANGES_REQUESTED** [review]

File: variables.tf (hunk 4 above)
Comment (swiftnickmoj) [reply variables.tf]: Added the line below to terratest:

  dns_resource_group_name      = var.dns_resource_group_name

**Review by Abhivan — APPROVED** [review]

**Comment by swiftnickmoj** [comment]
All issues remediated. P/R ready to be re-reviewed/re-examined

**Pull request merged**